plugin/src/main/kotlin/org/jlleitschuh/gradle/ktlint/BaseKtlintCheckTask.kt

```diff
@@ -104,39 +105,54 @@ abstract class BaseKtlintCheckTask(
+    @OutputFiles
+    private fun ktlintArgsFile() = newFileProperty(objectFactory, projectLayout).apply {
+        set(
+            project.layout.buildDirectory.file(
+                project.provider {
+                    "ktlint/${this@BaseKtlintCheckTask.name}.args"
+                }
+            )
+        )
+    }
+
+    abstract fun additionalConfig(): (PrintWriter) -> Unit
 
     private fun generateJavaExecSpec(
-        additionalConfig: (JavaExecSpec) -> Unit
+        additionalConfig: (PrintWriter) -> Unit
     ): (JavaExecSpec) -> Unit = { javaExecSpec ->
         javaExecSpec.classpath = classpath
         javaExecSpec.main = resolveMainClassName(ktlintVersion.get())
-        javaExecSpec.args(getSource().toRelativeFilesList())
-        if (verbose.get()) {
-            javaExecSpec.args("--verbose")
-        }
-        if (debug.get()) {
-            javaExecSpec.args("--debug")
-        }
-        if (android.get()) {
-            javaExecSpec.args("--android")
-        }
-        if (outputToConsole.get()) {
-            javaExecSpec.args("--reporter=plain")
-        }
-        if (coloredOutput.get()) {
-            javaExecSpec.args("--color")
-        }
-        if (enableExperimentalRules.get()) {
-            javaExecSpec.args("--experimental")
+        javaExecSpec.isIgnoreExitValue = ignoreFailures.get()
+
+        val argsConfigFile = ktlintArgsFile().get().asFile
+        if (!argsConfigFile.exists()) {
+            argsConfigFile.parentFile.mkdirs()
+            argsConfigFile.createNewFile()
         }
+        val argsWriter = argsConfigFile.printWriter()
+        if (verbose.get()) argsWriter.println("--verbose")
+        if (debug.get()) argsWriter.println("--debug")
+        if (android.get()) argsWriter.println("--android")
+        if (outputToConsole.get()) argsWriter.println("--reporter=plain")
+        if (coloredOutput.get()) argsWriter.println("--color")
+        if (enableExperimentalRules.get()) argsWriter.println("--experimental")
         if (additionalEditorconfigFile.isPresent) {
-            javaExecSpec.args("--editorconfig=${additionalEditorconfigFile.get().asFile.absolutePath}")
+            argsWriter.println("--editorconfig=${additionalEditorconfigFile.get().asFile.absolutePath}")
         }
-        javaExecSpec.args(ruleSetsClasspath.files.map { "--ruleset=${it.absolutePath}" })
-        javaExecSpec.isIgnoreExitValue = ignoreFailures.get()
-        javaExecSpec.args(enabledReports.map { it.asArgument() })
-        additionalConfig(javaExecSpec)
+        ruleSetsClasspath
+            .files
+            .map { "--ruleset=${it.absolutePath}" }
+            .forEach { argsWriter.println(it) }
+        enabledReports
+            .map { it.asArgument() }
+            .forEach { argsWriter.println(it) }
+        getSource()
+            .toRelativeFilesList()
+            .forEach { argsWriter.println(it) }
+        additionalConfig(argsWriter)
+        argsWriter.close()
```

Comment:
Shouldn't this be in a `use` block so that if any of the above fail, we don't end up leaving a dangling file descriptor?